File: source/python/trie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end = False

def trie_insert(root, word):
    node = root
    for ch in word:
        if ch not in node.children:
            node.children[ch] = TrieNode()
        node = node.children[ch]
    node.is_end = True

def trie_search(root, word):
    node = root
    for ch in word:
        if ch not in node.children:
            return False
        node = node.children[ch]
    return node.is_end

def trie_starts_with(root, prefix):
    node = root
    for ch in prefix:
        if ch not in node.children:
            return False
        node = node.children[ch]
    return True

root = TrieNode()
for word in ["apple", "app", "application", "bat", "ball"]:
    trie_insert(root, word)

print(f'search "app": {str(trie_search(root, "app")).lower()}')
print(f'search "ap": {str(trie_search(root, "ap")).lower()}')
print(f'search "bat": {str(trie_search(root, "bat")).lower()}')
print(f'starts_with "app": {str(trie_starts_with(root, "app")).lower()}')
print(f'starts_with "ba": {str(trie_starts_with(root, "ba")).lower()}')
print(f'starts_with "cat": {str(trie_starts_with(root, "cat")).lower()}')

File: source/go/trie.go
package main

import "fmt"

type TrieNode struct {
	children map[rune]*TrieNode
	isEnd    bool
}

func newTrieNode() *TrieNode {
	return &TrieNode{children: make(map[rune]*TrieNode)}
}

func trieInsert(root *TrieNode, word string) {
	node := root
	for _, ch := range word {
		if node.children[ch] == nil {
			node.children[ch] = newTrieNode()
		}
		node = node.children[ch]
	}
	node.isEnd = true
}

func trieSearch(root *TrieNode, word string) bool {
	node := root
	for _, ch := range word {
		if node.children[ch] == nil {
			return false
		}
		node = node.children[ch]
	}
	return node.isEnd
}

func trieStartsWith(root *TrieNode, prefix string) bool {
	node := root
	for _, ch := range prefix {
		if node.children[ch] == nil {
			return false
		}
		node = node.children[ch]
	}
	return true
}

func main() {
	root := newTrieNode()
	words := []string{"apple", "app", "application", "bat", "ball"}
	for _, w := range words {
		trieInsert(root, w)
	}

	fmt.Printf("search \"app\": %v\n", trieSearch(root, "app"))
	fmt.Printf("search \"ap\": %v\n", trieSearch(root, "ap"))
	fmt.Printf("search \"bat\": %v\n", trieSearch(root, "bat"))
	fmt.Printf("starts_with \"app\": %v\n", trieStartsWith(root, "app"))
	fmt.Printf("starts_with \"ba\": %v\n", trieStartsWith(root, "ba"))
	fmt.Printf("starts_with \"cat\": %v\n", trieStartsWith(root, "cat"))
}
